pkg/kubelet/cm/dra/plugin/noderesources.go

```diff
@@ -0,0 +1,464 @@
+	// because eventually these list calls succeed.
+	// TODO: can we avoid these error log entries? Perhaps wait here?
+
+	// We could use an indexer on driver name, but that seems overkill.
+	informer := resourceinformers.NewFilteredNodeResourceSliceInformer(c.kubeClient, resyncPeriod, nil, func(options *metav1.ListOptions) {
+		options.FieldSelector = "nodeName=" + c.nodeName
+	})
+	c.sliceStore = informer.GetStore()
+	handler, err := informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj any) {
+			slice, ok := obj.(*resourceapi.NodeResourceSlice)
+			if !ok {
+				return
+			}
+			logger.V(5).Info("NodeResourceSlice add", "slice", klog.KObj(slice))
+			c.queue.Add(slice.DriverName)
+		},
+		UpdateFunc: func(old, new any) {
+			oldSlice, ok := old.(*resourceapi.NodeResourceSlice)
+			if !ok {
+				return
+			}
+			newSlice, ok := new.(*resourceapi.NodeResourceSlice)
+			if !ok {
+				return
+			}
+			if loggerV := logger.V(6); loggerV.Enabled() {
+				loggerV.Info("NodeResourceSlice update", "slice", klog.KObj(newSlice), "diff", cmp.Diff(oldSlice, newSlice))
+			} else {
+				logger.V(5).Info("NodeResourceSlice update", "slice", klog.KObj(newSlice))
+			}
+			c.queue.Add(newSlice.DriverName)
+		},
+		DeleteFunc: func(obj any) {
+			if tombstone, ok := obj.(cache.DeletedFinalStateUnknown); ok {
+				obj = tombstone.Obj
+			}
+			slice, ok := obj.(*resourceapi.NodeResourceSlice)
+			if !ok {
+				return
+			}
+			logger.V(5).Info("NodeResourceSlice delete", "slice", klog.KObj(slice))
+			c.queue.Add(slice.DriverName)
+		},
+	})
+	if err != nil {
+		logger.Error(err, "Registering event handler on the NodeResourceSlice informer failed, disabling resource monitoring")
+		return
+	}
+
+	// Start informer and wait for our cache to be populated.
+	c.wg.Add(1)
+	go func() {
+		defer c.wg.Done()
+		informer.Run(ctx.Done())
+	}()
+	for !handler.HasSynced() {
+		select {
+		case <-time.After(time.Second):
+		case <-ctx.Done():
+			return
+		}
+	}
+	logger.Info("NodeResourceSlice informer has synced")
+
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *nodeResourcesController) processNextWorkItem(ctx context.Context) bool {
+	key, shutdown := c.queue.Get()
+	if shutdown {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	driverName := key.(string)
+
+	// Panics are caught and treated like errors.
+	var err error
+	func() {
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("internal error: %v", r)
+			}
+		}()
+		err = c.sync(ctx, driverName)
+	}()
+
+	if err != nil {
+		// TODO: contextual logging in utilruntime
+		utilruntime.HandleError(fmt.Errorf("processing driver %v: %v", driverName, err))
+		c.queue.AddRateLimited(key)
+
+		// Return without removing the work item from the queue.
+		// It will be retried.
+		return true
+	}
+
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *nodeResourcesController) sync(ctx context.Context, driverName string) error {
```

Comment:
I am able to understand what all the code in this function is doing, but I find it a bit hard to follow. Could we simplify it by:

1. Introducing the following helper function:
```
func indexOfModel(models []*resourceapi.NodeResourceModel, model *resourceapi.NodeResourceModel) int {
	for index, m := range models {
		if apiequality.Semantic.DeepEqual(m, model) {
			return index
		}
	}
	return -1
}
```

2. Changing the loop that populates `storedResourceIndices` and `obsoleteSlices` to:
```
	for _, obj := range slices {
		slice := obj.(*resourceapi.NodeResourceSlice)
		if slice.DriverName != driverName {
			continue
		}
		
		index := indexOfModel(driverResources, &slice.NodeResourceModel)
		if index >= 0 {
			storedResourceIndices.Insert(index)
			continue
		}
		
		obsoleteSlices = append(obsoleteSlices, slice)
	}
```

3. Remove the `else` from the loop further on, and replace it with a continue in the ìf` statement (with some extra newlines for clarity as well):
```
	numObsoleteSlices := len(obsoleteSlices)
	for index, resource := range driverResources {
		if storedResources.Has(index) {
			continue
		}
		
		if numObsoleteSlices > 0 {
			slice := obsoleteSlices[numObsoleteSlices-1]
			numObsoleteSlices--
			slice = slice.DeepCopy()
			slice.NodeResourceModel = *resource
			logger.V(5).Info("Reusing existing node resource slice", "slice", klog.KObj(slice))
			if _, err := c.kubeClient.ResourceV1alpha2().NodeResourceSlices().Update(ctx, slice, metav1.UpdateOptions{}); err != nil {
				return fmt.Errorf("update node resource slice: %w", err)
			}
			continue
		}
		
		slice := &resourceapi.NodeResourceSlice{
			ObjectMeta: metav1.ObjectMeta{
				GenerateName: c.nodeName + "-" + driverName + "-",
				// TODO: node object as owner
			},
			NodeName:          c.nodeName,
			DriverName:        driverName,
			NodeResourceModel: *resource,
		}
		logger.V(5).Info("Creating new node resource slice", "slice", klog.KObj(slice))
		if _, err := c.kubeClient.ResourceV1alpha2().NodeResourceSlices().Create(ctx, slice, metav1.CreateOptions{}); err != nil {
			return fmt.Errorf("create node resource slice: %w", err)
		}
	}
```

Reply:
Yes, that's better. I also added a few more comment for each of these `if` checks.